script/reviewer.go

```diff
@@ -19,11 +30,88 @@ func (r *Reviewer) Review() {
 
 	r.application.Parse(r.gitHub.Issue)
 
-	if isTestingIssue() {
-		if r.application.IsValid() {
-			debugMessage("Application has no problems")
+	status := r.getStatus()
+	r.updateLabels(status)
+	r.createComment(status)
+}
+
+func (r *Reviewer) getStatus() Status {
+	if *r.gitHub.Issue.State == "closed" {
+		return Closed
+	} else if r.gitHub.IssueHasLabel(LabelStatusApproved) {
+		return Approved
+	} else if r.gitHub.IssueHasLabel(LabelStatusReviewing) {
+		return Reviewing
+	} else if r.gitHub.IssueHasLabel(LabelStatusInvalid) {
+		return Invalid
+	} else {
+		return New
+	}
+}
+
+func (r *Reviewer) createComment(status Status) {
+	title := ""
+	body := ""
+	details := fmt.Sprintf("<details>\n<summary>Application data...</summary>\n\n```json\n%s\n```\n</details>", r.application.GetData())
```

Comment:
On a second look here... you're making a GitHub comment using untrusted input from our users. What's to prevent them from adding their own HTML or XSS attacks here? Which user is making these comments and could be hijacked? I don't think that GitHub supports slash commands like GitLab (where they could force your bot to `/approve` the issue), but am I missing something there?

Reply:
GitHub has extremely [strict markdown sanitization rules](https://github.github.com/gfm/) (otherwise any Issue/PR/Readme could XSS you). GH doesn't directly support any slash commands _that perform actions_, however, if we implemented chat/slash commands with GH Actions down the road, yes this could be a problem.

Reply:
Great callout! By and large `json.MarshalIndent`, used by the `GetData` method, escapes `<`, `>`, `&` and produces valid JSON, reducing the chances of breaking out of the printed JSON.

That said, I've hardened this even further by parsing all input data as Markdown into HTML, then retrieving only the text content of the HTML. This will help:

1. Protect us against Markdown syntax being used to break out of the JSON. After all, the JSON is itself being printed inside of a Markdown code block in the issue comment, so a user might have tried to use three backticks to end the code block.
2. Protect consumers of this data down the road are better protected if they directly render one of the JSON values (like the project description).

> Which user is making these comments and could be hijacked?

We are employing a bot account to make these comments. A PAT for that account is stored as a secret on this repository and passed to the script when the workflow is invoked. Absent this PAT the script will bail.

> I don't think that GitHub supports slash commands like GitLab (where they could force your bot to /approve the issue), but am I missing something there?

On the approval side, the workflow to approve and commit the data to the repository is only invoked when the `status: approved` label is added to the issue, and labels can only be added by a collaborator on the repository. Only AgileBits employees can be collaborators on the repository.